design-documents/frontend/server-side-rendering.md

```diff
@@ -0,0 +1,129 @@
+- User may see some content before it's interactive.
+- Difficult or tedious to implement.
+- May require additional configuration and refactoring.
+- Some features [not yet supported](#not-supported-yet).
+
+## How it works
+
+The basic idea is that we don't want the web browser (client) to be responsible for rendering content via JavaScript, because JavaScript is notoriously slow, especially at rendering DOM and especially on older or mobile devices. Unnecessary rendering also consumes more energy, which has the potential to suck a mobile device's battery dry. :battery:
+
+If we leverage the HTTP server to do as much heavy lifting as we can, some of this heavy lifting (non-sensitive pages) can be cached as well, resulting in a faster, better user experience.
+
+### Without SSR
+
+1. A web browser (client) requests a URL for a particular route (e.g., `/`).
+1. The HTTP server runs the appropriate code for that route.
+   - HTML is partially generated with empty containers that need to be filled and rendered by JavaScript after the initial page load (bad for SEO).
+1. The web browser (client) receives and processes the HTML, but the user may see some empty content that hasn't been filled in yet.
+1. JavaScript is loaded and processed, including the React library.
+1. Ideally, a single GraphQL API request is made to collect the data required to fill in the empty containers with content.
+1. React renders the final HTML into the above-mentioned empty containers by combining the data received with the HTML defined in the React components.
+1. Once content is seen, it becomes immediately interactive.
+
+### With SSR
+
+1. A web browser (client) requests a URL for a particular route (e.g., `/`).
+1. The [JavaScript-capable HTTP server](#requirements) runs the appropriate code for that route.
+   - Ideally, a single GraphQL API request is made from this HTTP server, collecting all the data that's required to render the page.
+   - HTML is fully generated [via React methods](https://reactjs.org/docs/react-dom-server.html) with no empty containers. This HTML can be [streamed](https://reactjs.org/docs/react-dom-server.html#rendertonodestream) and cached.
+1. The web browser (client) receives and processes the HTML.
+   - No additional HTML needs to be rendered into containers. As a result, the user perceives faster performance.
+   - At this point, some content might appear to be interactive (e.g., buttons), but won't actually work yet.
+1. JavaScript is loaded and processed, including the React library.
+1. React [hydrates](https://reactjs.org/docs/react-dom.html#hydrate) any containers whose HTML contents were rendered by [ReactDOMServer](https://reactjs.org/docs/react-dom-server.html) in step 2 (above). Nothing needs to be re-rendered here.
+1. The page is now fully interactive.
+1. Subsequent API requests are made directly to the GraphQL API.
+
+## Requirements
+
+- A JavaScript-capable HTTP server, typically a [Node.js][] server (e.g., [Express][]).
```

Comment:
Should clarify this: you don't need a JavaScript-capable HTTP server - you just need your web server to have access to _something_ that let's it execute JavaScript. That does not need to live within the HTTP server itself. Small but important distinction

Reply:
Yes, technically, you are correct. I really don't want to point people down that path though, so I'll just stick to my "non-golden path" and "not recommended" piece in the FAQ section.